Source/Tests/Scene/Prefab.cpp

```diff
@@ -0,0 +1,148 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR rhs DEALINGS IN
+// THE SOFTWARE.
+//
+
+#include "../CommonUtils.h"
+#include "../SceneUtils.h"
+#include "Urho3D/Graphics/StaticModel.h"
+
+#include <Urho3D/IO/MemoryBuffer.h>
+#include <Urho3D/Resource/ResourceCache.h>
+#include <Urho3D/Resource/ResourceEvents.h>
+#include <Urho3D/Physics/Constraint.h>
+#include <Urho3D/Physics/RigidBody.h>
+
+#include <Urho3D/Scene/PrefabReference.h>
+
+TEST_CASE("Prefab reference")
+{
+    auto context = Tests::GetOrCreateContext(Tests::CreateCompleteContext);
+    auto cache = context->GetSubsystem<ResourceCache>();
+    auto scene = MakeShared<Scene>(context);
+
+    auto node0 = scene->CreateChild();
+    auto node1 = scene->CreateChild();
+
+    auto xmlFile = new XMLFile(context);
+    xmlFile->SetName("Objects/Obj0.xml");
+    auto fileText = MemoryBuffer("<node><component type=\"StaticModel\"/></node>");
+    REQUIRE(xmlFile->Load(fileText));
+    cache->AddManualResource(xmlFile);
+
+    SharedPtr<PrefabReference> prefabRef{node0->CreateComponent<PrefabReference>(CreateMode::REPLICATED)};
+    prefabRef->SetPrefab(xmlFile);
+
+    // Setting prefab to enabled node makes component to create temporary node attached to the component's node.
+    // Make it shared ptr to ensure that new node won't be allocated at the same address.
+    SharedPtr<Node> prefabRoot{prefabRef->GetRootNode()};
+    REQUIRE(prefabRoot != nullptr);
+    CHECK(prefabRoot->IsTemporary());
+    CHECK(prefabRoot->GetParent() == node0);
+    CHECK(prefabRoot->GetNumChildren() == 0);
+
+    // Component should preserve the node but detach it from the parent
+    prefabRef->Remove();
+    REQUIRE(prefabRef->GetRootNode() == prefabRoot);
+    CHECK(prefabRoot->GetParent() == nullptr);
+
+    // Moving component to another node makes prefab root attached
+    node1->AddComponent(prefabRef, prefabRef->GetID(), CreateMode::REPLICATED);
+    CHECK(prefabRoot->GetParent() == node1);
+
+    // Reload the prefab on file change
+    auto fileText2 = MemoryBuffer("<node><component type=\"StaticModel\"/><node/></node>");
+    xmlFile->Load(fileText2);
+    {
+        using namespace ReloadFinished;
+        VariantMap data;
+        xmlFile->SendEvent(E_RELOADFINISHED, data);
+    }
+    CHECK(prefabRef->GetRootNode() != prefabRoot);
+    prefabRoot = prefabRef->GetRootNode();
+    CHECK(prefabRoot->GetNumChildren() == 1);
+
+    prefabRef->Inline();
+    CHECK(prefabRef->GetNode() == nullptr);
+    CHECK(!prefabRoot->IsTemporary());
+}
+
+TEST_CASE("Prefab with node reference")
+{
+    auto context = Tests::GetOrCreateContext(Tests::CreateCompleteContext);
+    auto cache = context->GetSubsystem<ResourceCache>();
+    auto scene = MakeShared<Scene>(context);
+
+    auto node0 = scene->CreateChild();
+    auto node1 = scene->CreateChild();
+
+    auto xmlFile = new XMLFile(context);
+    xmlFile->SetName("Objects/Obj1.xml");
+    auto fileText = MemoryBuffer("<node id=\"1\">"
```

Comment:
Any reason to use text instead of building xml with functions?
I don't insist tho.
But if you want to use text, please indent it at least